### PR TITLE
Make kubectl bash-completion namespace and resource alias aware

### PR DESCRIPTION
**Pull request by sttts**
kubectl-completion
- filter resource listing by `--namespace` flag given before in the command line

``` bash
$ kubectl get pod --namespace=kube-system <tab><tab>
kube-dns-v9-2wuzj   kube-dns-v9-llqxa
```
- add completion of `--namespace`

``` bash
$ kubectl get pod --namespace=<tab><tab>
[*]         default      ingress      kube-system
```
- add support for plural nouns and aliases like `rc`

### CODE REVIEW

File: pkg/kubectl/cmd/cmd.go [hunk 1]
```diff
@@ -30,12 +30,47 @@ import (
 
 const (
 	bash_completion_func = `# call kubectl get $1,
+__kubectl_namespace_flag()
+{
+    local ret two_word_ns
+    ret=""
+    two_word_ns=false
+    for w in ${words[@]}; do
```
Comment (eparis) [inline pkg/kubectl/cmd/cmd.go]: ${words[@]} should be in quotes.

Comment (sttts) [inline pkg/kubectl/cmd/cmd.go]: Fixed.

File: contrib/completions/bash/kubectl [hunk 2]
```diff
@@ -135,6 +166,8 @@ __handle_noun()
 
     if __contains_word "${words[c]}" "${must_have_one_noun[@]}"; then
         must_have_one_noun=()
+    elif __contains_word "${words[c]%!s(MISSING)}" "${must_have_one_noun[@]}"; then
```
Comment (eparis) [inline contrib/completions/bash/kubectl]: `%!s(MISSING)` ?

Comment (sttts) [inline contrib/completions/bash/kubectl]: Bad escaping of bash code in golang string in fact: https://github.com/spf13/cobra/pull/255/files#diff-3db5cab63dbb32fff4264f5449fd2904R192

**Comment by sttts** [comment]
hack/update-godep-licenses.sh doesn't like me.

**Comment by sttts** [comment]
bash4'ism in that script.

**Comment by sttts** [comment]
ptal @eparis. Everything green.

**Comment by sttts** [comment]
- namespaced all zsh function to avoid conflicts with git completion and others
- updated godeps to https://github.com/spf13/cobra/pull/257
- added noun aliases
- added proper compgen filtering for zsh

**Comment by sttts** [comment]
Rebased. But because it's based on unmerged spf13/cobra#257, it's still red due to invalid godeps.

File: contrib/completions/zsh/kubectl [hunk 3]
```diff
@@ -0,0 +1,214 @@
+#!/bin/zsh
```
Comment (eparis) [inline contrib/completions/zsh/kubectl]: This file was just created my hand right? Not generated in any way? Should it be a separate PR?

Comment (sttts) [inline contrib/completions/zsh/kubectl]: Yes, it's just a collection of functions that zsh's bash-completion emulation does not provide or which are incomplete. Will create an independent PR.

Comment (sttts) [inline contrib/completions/zsh/kubectl]: Moved into https://github.com/kubernetes/kubernetes/pull/23797

**Comment by sttts** [comment]
Updated to latest cobra master with shellcheck fixes.

**Comment by sttts** [comment]
@k8s-bot test this please flake: #21367
